src/vsgi-request.vala

```diff
@@ -389,9 +396,13 @@ namespace VSGI {
 				_headers = new Soup.MessageHeaders (Soup.MessageHeadersType.REQUEST);
 			}
 			if (_query == null) {
-				_query = uri.get_query () == null ? null : (HashTable<string,string>) Soup.Form.decode (uri.get_query ());
+				_query = _uri.get_query () == null ? null : (HashTable<string,string>) Soup.Form.decode (_uri.get_query ());
 			} else {
-				_uri.set_query_from_form (_query);
+				_uri_query = Soup.Form.encode (_query);
+			}
+			if (_uri == null) {
+				// TODO
+				_uri = Uri.build (UriFlags.ENCODED_PATH, _uri_scheme, _uri_userinfo, _uri_host, _uri_port, _uri_path_encoded, _uri_query, null);
```

Comment:
This will need a little bit more testing for various scenarios where the URI is set explicitly, or via CGI environment.